### PR TITLE
Refactor pagecache evict

### PR DESCRIPTION
**Pull request by sanderssj**
This change addresses issue #1253 by moving evicted pages to a free list when no longer used, but leaving their meta in the node's tree. When the page is accessed again, memory is reallocated. This change also fixes a missing reserve_refcount in map_pages.

### CODE REVIEW

**Comment by sanderssj** [comment]
The io_uring test sometimes fails with a page fault, suggesting there's still some refcount mishandling. I need to find out what causing that problem before merging to master.

**Review by wjhun — COMMENTED** [review]
I thought that we should be able to re-instate the refcount_release in pagecache_write_sg_finish (line 518). But when I tried it, I consistently get a failure in the io_uring test:

```
assigned: 10.0.2.15                                                                                                                                                                                                                            
ffffffff8000a281        (pagecache_page_free + 0000000000000151/00000000000001ab)                                                                                                                                                              
ffffffff8000a993        (pagecache_write_sg_finish + 00000000000001d3/0000000000000c26)                                                                                                                                                        
ffffffff8001f980        (merge_join + 0000000000000040/0000000000000098)                                                                                                                                                                       
ffffffff8001f980        (merge_join + 0000000000000040/0000000000000098)                                                                                                                                                                       
ffffffff80066dc5        (virtio_scsi_io_done + 0000000000000025/00000000000000a2)                                                                                                                                                              
ffffffff80066577        (virtio_scsi_request_complete + 0000000000000027/000000000000008b)                                                                                                                                                     
ffffffff80068151        (virtqueue_service_vqmsgs + 00000000000001a1/00000000000001b8)                                                                                                                                                         
ffffffff80010362        (runloop_internal + 0000000000000292/000000000000032b)                                                                                                                                                                 
assertion pp->write_count == 0 failed in /share/src/nanovms/justin-nanos/src/kernel/pagecache.c: pagecache_page_free() on line 367; halt
```

Is this the same failure that you saw occur sporadically last week? (Edit: the remaining tests appear to pass without issue.)

File: src/kernel/pagecache.c [hunk 1]
```diff
@@ -616,49 +614,86 @@ closure_function(1, 3, void, pagecache_write_sg,
...
+    vector v;
+    u64 evicted = 0;
+    u64 lastevicted = 0;
 
-    /* We could avoid taking both locks here if we keep drained page
-       objects around (which incidentally could be useful to keep
-       refault data). */
+    if ((v = allocate_vector(pc->h, DRAIN_ITER_MAX)) == INVALID_ADDRESS)
+        return 0;
+    while (evicted < pages) {
```
Comment (wjhun) [inline src/kernel/pagecache.c:680]: Pages is being decremented while evicted is accumulating, so this should either change to pages > 0, or perhaps better yet, remove the pages decrement below and use MIN(pages - evicted, DRAIN_ITER_MAX). Also, keeping lastevicted isn't really necessary - could just check the return value of evict_pages_locked, break if 0, else add to evicted.

File: src/kernel/pagecache.c [hunk 2]
```diff
@@ -453,7 +478,7 @@ closure_function(5, 1, void, pagecache_write_sg_finish,
 
     pagecache_lock_node(pn);
     pagecache_page pp = page_lookup_nodelocked(pn, pi);
-    if (bound(complete)) {
+    if (!is_ok(s) || bound(complete)) {
```
Comment (wjhun) [inline src/kernel/pagecache.c:481]: Really minor, but I would update the debug below to indicate whether the error was on write (complete true) or read. Yet I would still record a read error in pv->write_error, for it is an error for the cache write operation as a whole.

**Review by wjhun — APPROVED** [review]
Just one issue with the loop in pagecache_drain, but otherwise looks good.

File: src/kernel/pagecache.c (hunk 1 above)
Comment (sanderssj) [reply src/kernel/pagecache.c:679]: Nice catch! I corrected and cleaned it up.

**Pull request merged**